File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        cl = set(['(', '{' , '['])
        mp = {
            ']':'[',
        ')':'(', 
        '}':'{'
        }
        st = []
        for i in s:
            if i in cl:
                st.append(i)
            elif i in mp:
                if len(st)!=0 and (st[-1] == mp[i]):
                   
                    st.pop()
                else:
                    return False
        return len(st)==0

        

File: 0020-valid-parentheses/0020-valid-parentheses.cs
public class Solution {
    public bool IsValid(string s) {
        //  HashSet<char> cl = new HashSet<char> { '(', '{', '[' };
        //  Dictionary<char , char> mp = new Dictionary<char , char> {
        //     { ')', '(' },
        //     { '}', '{' },
        //     { ']', '[' }
        // };
        //  Stack<char> stack = new Stack<char>();
        //  foreach(char i in s.ToCharArray()){
        //         if ( cl.Contains(i)){
        //             stack.Push(i);
        //         }
        //         else if (mp.ContainsKey(i) && stack.Count>0)
        //         {
        //             if (stack.Peek()==mp[i]){
        //                 stack.Pop();
        //             }
        //         else
        //         {
        //                 return false;
        //         }
                    
        //     }
        //  }
        //  return stack.Count==0;

        var stack = new Stack<char>();
        var matchingBrackets = new Dictionary<char, char> {
            { ')', '(' },
            { '}', '{' },
            { ']', '[' }
        };

        foreach (char ch in s) {
            if (matchingBrackets.ContainsValue(ch)) {
                stack.Push(ch);
            } else if (matchingBrackets.ContainsKey(ch)) {
                if (stack.Count == 0 || stack.Pop() != matchingBrackets[ch]) {
                    return false;
                }
            }
        }

        return stack.Count == 0;

    }
}